README.md
[![NPM version](https://badge.fury.io/js/cdk-kaniko.svg)](https://badge.fury.io/js/cdk-kaniko)
[![PyPI version](https://badge.fury.io/py/cdk-kaniko.svg)](https://badge.fury.io/py/cdk-kaniko)
[![Release](https://github.com/pahud/cdk-kaniko/actions/workflows/release.yml/badge.svg)](https://github.com/pahud/cdk-kaniko/actions/workflows/release.yml)

# `cdk-kaniko`

Build images with `kanilo` in **AWS Fargate**

# About

`cdk-kaniko` is a CDK construct library that allows you to build images with [**kaniko**](https://github.com/GoogleContainerTools/kaniko) in **AWS Fargate**. Inspired from the blog post - [Building container images on Amazon ECS on AWS Fargate](https://aws.amazon.com/tw/blogs/containers/building-container-images-on-amazon-ecs-on-aws-fargate/) by _Re Alvarez-Parmar_ and _Olly Pomeroy_, this library aims to abstract away all the infrastructure provisioning and configuration with minimal IAM policies required and allow you to focus on the high level CDK constructs. Under the covers, `cdk-kaniko` leverages the [cdk-fargate-run-task](https://github.com/pahud/cdk-fargate-run-task) so you can build the image just once or schedule the building periodically.

# Sample

```ts
const app = new cdk.App();

const stack = new cdk.Stack(app, 'my-stack-dev');

const kaniko = new Kaniko(stack, 'KanikoDemo', {
  context: 'git://github.com/pahud/vscode.git',
  contextSubPath: './.devcontainer',
});

// build it once
kaniko.buildImage('once');

// schedule the build every day 0:00AM
kaniko.buildImage('everyday', Schedule.cron({
  minute: '0',
  hour: '0',
}));
```

# fargate spot support

Use `fargateSpot` to enable the `FARGATE_SPOT` capacity provider to provision the fargate tasks.

```ts
new Kaniko(stack, 'KanikoDemo', {
  context,
  contextSubPath,
  fargateSpot: true,
});
```

# Note

Please note the image building could take some minutes depending on the complexity of the provided `Dockerfile`. On deployment completed, you can check and tail the **AWS Fargate** task logs from the **AWS CloudWatch Logs** to view all the build output.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.4](https://github.com/pahud/cdk-kaniko/compare/v0.2.3...v0.2.4) (2021-04-05)

### [0.2.3](https://github.com/pahud/cdk-kaniko/compare/v0.2.2...v0.2.3) (2021-04-05)

### [0.2.2](https://github.com/pahud/cdk-kaniko/compare/v0.2.1...v0.2.2) (2021-04-04)


### Bug Fixes

* add-run-task-dependency ([#8](https://github.com/pahud/cdk-kaniko/issues/8)) ([6222dca](https://github.com/pahud/cdk-kaniko/commit/6222dcac280127df9ab1ec5e89822bfebd17ca82)), closes [#7](https://github.com/pahud/cdk-kaniko/issues/7)

### [0.2.1](https://github.com/pahud/cdk-kaniko/compare/v0.2.0...v0.2.1) (2021-04-03)

## [0.2.0](https://github.com/pahud/cdk-kaniko/compare/v0.1.3...v0.2.0) (2021-04-02)


### Features

* support fargate spot ([#4](https://github.com/pahud/cdk-kaniko/issues/4)) ([3ab51b5](https://github.com/pahud/cdk-kaniko/commit/3ab51b5931bba47dce14cfc84f965ead7c0f572e))

### [0.1.3](https://github.com/pahud/cdk-kaniko/compare/v0.1.2...v0.1.3) (2021-04-02)

### [0.1.2](https://github.com/pahud/cdk-kaniko/compare/v0.1.1...v0.1.2) (2021-04-02)

### [0.1.1](https://github.com/pahud/cdk-kaniko/compare/v0.1.0...v0.1.1) (2021-04-02)

## 0.1.0 (2021-04-02)


test/integ.default.test.ts
import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import * as cdk from '@aws-cdk/core';
import { Kaniko } from '../src';

test('minimal usage', () => {
  // GIVEN
  const app = new cdk.App();

  const env = {
    region: process.env.CDK_DEFAULT_REGION || 'us-east-1',
    account: process.env.CDK_DEFAULT_ACCOUNT || '123456789',
  };

  const stack = new cdk.Stack(app, 'my-stack-dev', { env });
  // WHEN
  const kaniko = new Kaniko(stack, 'KanikoDemo', {
    context: 'git://github.com/pahud/vscode.git',
    contextSubPath: './.devcontainer',
  });

  kaniko.buildImage('once');

  // THEN
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();

  expect(stack).toHaveResourceLike('AWS::ECS::TaskDefinition', {
    ContainerDefinitions: [
      {
        Command: [
          '--context',
          'git://github.com/pahud/vscode.git',
          '--context-sub-path',
          './.devcontainer',
          '--dockerfile',
          'Dockerfile',
          '--destination',
          {
            'Fn::Join': [
              '',
              [
                {
                  'Fn::Select': [
                    4,
                    {
                      'Fn::Split': [
                        ':',
                        {
                          'Fn::GetAtt': [
                            'KanikoDemoRepoE4811266',
                            'Arn',
                          ],
                        },
                      ],
                    },
                  ],
                },
                '.dkr.ecr.',
                {
                  'Fn::Select': [
                    3,
                    {
                      'Fn::Split': [
                        ':',
                        {
                          'Fn::GetAtt': [
                            'KanikoDemoRepoE4811266',
                            'Arn',
                          ],
                        },
                      ],
                    },
                  ],
                },
                '.',
                {
                  Ref: 'AWS::URLSuffix',
                },
                '/',
                {
                  Ref: 'KanikoDemoRepoE4811266',
                },
              ],
            ],
          },
          '--force',
        ],
        Essential: true,
        Image: {
          'Fn::Join': [
            '',
            [
              '123456789.dkr.ecr.us-east-1.',
              {
                Ref: 'AWS::URLSuffix',
              },
              '/aws-cdk/assets:8b0a26aabbdd052682e08242ae76cd903c22e3a8356ea0f5c15e536b6a7c99ab',
            ],
          ],
        },
        LogConfiguration: {
          LogDriver: 'awslogs',
          Options: {
            'awslogs-group': {
              Ref: 'KanikoDemoBuildImageTaskkanikoLogGroup4C2A8704',
            },
            'awslogs-stream-prefix': 'kaniko',
            'awslogs-region': 'us-east-1',
          },
        },
        Name: 'kaniko',
      },
    ],
    Cpu: '512',
    ExecutionRoleArn: {
      'Fn::GetAtt': [
        'KanikoDemoBuildImageTaskExecutionRoleB24935D9',
        'Arn',
      ],
    },
    Family: 'mystackdevKanikoDemoBuildImageTaskB3BE4585',
    Memory: '1024',
    NetworkMode: 'awsvpc',
    RequiresCompatibilities: [
      'FARGATE',
    ],
    TaskRoleArn: {
      'Fn::GetAtt': [
        'KanikoDemoBuildImageTaskTaskRole84C9326D',
        'Arn',
      ],
    },
  });
});


.projenrc.js
const { AwsCdkConstructLibrary } = require('projen');
const { Automation } = require('projen-automate-it');

const AUTOMATION_TOKEN = 'PROJEN_GITHUB_TOKEN';

const project = new AwsCdkConstructLibrary({
  author: 'Pahud Hsieh',
  authorAddress: 'pahudnet@gmail.com',
  cdkVersion: '1.96.0',
  defaultReleaseBranch: 'main',
  jsiiFqn: 'projen.AwsCdkConstructLibrary',
  name: 'cdk-kaniko',
  description: 'CDK construct library that allows you to build docker images with kaniko in AWS Fargate',
  repositoryUrl: 'https://github.com/pahud/cdk-kaniko.git',
  cdkDependencies: [
    '@aws-cdk/core',
    '@aws-cdk/aws-ec2',
    '@aws-cdk/aws-ecr',
    '@aws-cdk/aws-ecs',
    '@aws-cdk/aws-events',
  ],
  deps: ['cdk-fargate-run-task'],
  peerDeps: ['cdk-fargate-run-task'],
  dependabot: false,
  publishToPypi: {
    distName: 'cdk-kaniko',
    module: 'cdk_kaniko',
  },
  devDeps: ['projen-automate-it'],
  keywords: [
    'kaniko',
    'cdk',
    'fargate',
    'aws',
  ],
});

const automation = new Automation(project, {
  automationToken: AUTOMATION_TOKEN,
});

automation.projenYarnUpgrade();

const common_exclude = ['cdk.out', 'cdk.context.json', 'images', 'yarn-error.log'];
project.npmignore.exclude(...common_exclude);
project.gitignore.exclude(...common_exclude);

project.synth();



src/integ.default.ts
import { Schedule } from '@aws-cdk/aws-events';
import * as cdk from '@aws-cdk/core';
import { Kaniko } from './';

export class IntegTesting {
  readonly stack: cdk.Stack[];
  constructor() {
    const app = new cdk.App();

    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new cdk.Stack(app, 'my-stack-dev', { env });

    const kaniko = new Kaniko(stack, 'KanikoDemo', {
      context: 'git://github.com/pahud/vscode.git',
      contextSubPath: './.devcontainer',
      fargateSpot: true,
    });

    // build it once
    kaniko.buildImage('once');

    // schedule the build every day 0:00AM
    kaniko.buildImage('everyday', Schedule.cron({
      minute: '0',
      hour: '0',
    }));

    this.stack = [stack];
  }
}

new IntegTesting();


src/kaniko.ts
import * as path from 'path';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecr from '@aws-cdk/aws-ecr';
import * as ecs from '@aws-cdk/aws-ecs';
import { Schedule } from '@aws-cdk/aws-events';
import * as cdk from '@aws-cdk/core';
import { RunTask } from 'cdk-fargate-run-task';

export interface KanikoProps {
  /**
   * Kaniko build context.
   * @see https://github.com/GoogleContainerTools/kaniko#kaniko-build-contexts
   */
  readonly context: string;
  /**
   * The target ECR repository
   * @default - create a new ECR private repository
   */
  readonly destinationRepository?: ecr.IRepository;
  /**
   * The context sub path
   *
   * @defautl - current directory
   */
  readonly contextSubPath?: string;
  /**
   * The Dockerfile for the image building
   *
   * @default Dockerfile
   */
  readonly dockerfile?: string;
  /**
   * Use FARGATE_SPOT capacity provider.
   */
  readonly fargateSpot?: boolean;
}

export class Kaniko extends cdk.Construct {
  readonly destinationRepository: ecr.IRepository
  readonly cluster: ecs.ICluster;
  readonly task: ecs.FargateTaskDefinition;
  readonly vpc: ec2.IVpc;
  private fargateSpot: boolean;
  constructor(scope: cdk.Construct, id: string, props: KanikoProps) {
    super(scope, id);

    this.fargateSpot = props.fargateSpot ?? false;

    this.vpc = getOrCreateVpc(this);
    this.cluster = new ecs.Cluster(this, 'Cluster', {
      vpc: this.vpc,
      capacityProviders: ['FARGATE', 'FARGATE_SPOT'],
    });
    this.destinationRepository = props.destinationRepository ?? this._createDestinationRepository();

    const executorImage = ecs.ContainerImage.fromAsset(path.join(__dirname, '../docker.d'));

    this.task = new ecs.FargateTaskDefinition(this, 'BuildImageTask', {
      cpu: 512,
      memoryLimitMiB: 1024,
    });
    this.task.addContainer('kaniko', {
      image: executorImage,
      command: [
        '--context', props.context,
        '--context-sub-path', props.contextSubPath ?? './',
        '--dockerfile', props.dockerfile ?? 'Dockerfile',
        '--destination', this.destinationRepository.repositoryUri,
        '--force',
      ],
      logging: new ecs.AwsLogDriver({ streamPrefix: 'kaniko' }),
    });

    this.destinationRepository.grantPullPush(this.task.taskRole);

    new cdk.CfnOutput(this, 'Repository', {
      value: this.destinationRepository.repositoryName,
    });


  }
  private _createDestinationRepository(): ecr.Repository {
    return new ecr.Repository(this, 'Repo', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
  }
  /**
   * Build the image with kaniko.
   * @param schedule The schedule to repeatedly build the image
   */
  public buildImage(id: string, schedule?: Schedule) {
    // run it just once
    const newRunTask = new RunTask(this, `BuildImage${id}`, {
      task: this.task,
      cluster: this.cluster,
      schedule,
      capacityProviderStrategy: this.fargateSpot ? [
        {
          capacityProvider: 'FARGATE_SPOT',
          weight: 1,
        },
      ] : undefined,
    });
    // if vpc is a new resource in this stack, run task job will add dependency on vpc created.
    if ((this.node.tryFindChild('Vpc') as ec2.Vpc).node.children.find(c => (c as cdk.CfnResource).cfnResourceType === 'AWS::EC2::VPC')) {
      newRunTask.node.addDependency(this.vpc);
    };
  }
}

function getOrCreateVpc(scope: cdk.Construct): ec2.IVpc {
  // use an existing vpc or create a new one
  return scope.node.tryGetContext('use_default_vpc') === '1'
    || process.env.CDK_USE_DEFAULT_VPC === '1' ? ec2.Vpc.fromLookup(scope, 'Vpc', { isDefault: true }) :
    scope.node.tryGetContext('use_vpc_id') ?
      ec2.Vpc.fromLookup(scope, 'Vpc', { vpcId: scope.node.tryGetContext('use_vpc_id') }) :
      new ec2.Vpc(scope, 'Vpc', { maxAzs: 3, natGateways: 1 });
}
